### ISSUE TITLE
JavaScript error: , line 0: uncaught exception: undefined

### ISSUE THREAD

**Issue by ishan-biztech** [mozilla/geckodriver#1792]
## System

* Version: geckodriver 0.27.0 (7b8c4f32cdde 2020-07-28 18:16 +0000)
* Platform: CentOS Linux + 7 (Core)
* Firefox: Mozilla Firefox 76.0
* Selenium: selenium + 3.141.0

## Testcase

```python
cert_verification_url = 'https://www.psacard.com/cert/46410267'
cert_verification_title = 'Certification Verification 46410267'
driver.get(cert_verification_url)

WebDriverWait(driver, 10).until(EC.title_is(cert_verification_title))

psa_cert_images_el = WebDriverWait(driver, 20).until(
    EC.presence_of_element_located((By.ID, 'psaCertImages'))
)
main_content_el = driver.find_element_by_id('mainContent');

# Here text returns as empty
print(main_content_el.text);
```

Above code works perfectly fine when using command line to run this script. But when we call it from our PHP script using `shell_exec` function it returns error "JavaScript error: , line 0: uncaught exception: undefined" in `geckodriver.log` file.

Below is output from `geckodriver.log` file:

```bash
*** You are running in headless mode.
1602673387764	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602673388133	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602673388133	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602673388133	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602673388133	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602673389322	Marionette	INFO	Listening on port 33215
1602673389418	Marionette	WARN	TLS certificate errors will be ignored for this session
1602673395495	Marionette	INFO	Stopped listening on port 33215
JavaScript error: , line 0: uncaught exception: undefined
1602673656247	geckodriver	INFO	Listening on 127.0.0.1:48789
1602673656251	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-headless" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofileLAHLLE"
*** You are running in headless mode.
1602673656436	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602673656730	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602673656730	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602673656730	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602673656730	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602673657745	Marionette	INFO	Listening on port 34505
1602673657756	Marionette	WARN	TLS certificate errors will be ignored for this session
1602673663926	Marionette	INFO	Stopped listening on port 34505
JavaScript error: , line 0: uncaught exception: undefined
1602674118527	geckodriver	INFO	Listening on 127.0.0.1:45765
1602674118531	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-headless" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofileBnibpw"
*** You are running in headless mode.
1602674118708	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602674119016	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602674119016	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602674119016	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602674119016	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602674120076	Marionette	INFO	Listening on port 41367
1602674120137	Marionette	WARN	TLS certificate errors will be ignored for this session
1602674126333	Marionette	INFO	Stopped listening on port 41367
JavaScript error: , line 0: uncaught exception: undefined
1602674149895	geckodriver	INFO	Listening on 127.0.0.1:60677
1602674149899	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-headless" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofilexoKQ1h"
*** You are running in headless mode.
1602674150072	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602674150354	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602674150355	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602674150355	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602674150355	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602674151273	Marionette	INFO	Listening on port 44199
1602674151306	Marionette	WARN	TLS certificate errors will be ignored for this session
1602674157278	Marionette	INFO	Stopped listening on port 44199
1602674335992	geckodriver	INFO	Listening on 127.0.0.1:41161
1602674335997	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-headless" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofileNC4pYm"
*** You are running in headless mode.
1602674336188	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602674336519	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602674336519	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602674336519	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602674336519	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602674337547	Marionette	INFO	Listening on port 32793
1602674337601	Marionette	WARN	TLS certificate errors will be ignored for this session
1602674344557	Marionette	INFO	Stopped listening on port 32793
JavaScript error: , line 0: uncaught exception: undefined
1602674402670	geckodriver	INFO	Listening on 127.0.0.1:41147
1602674402680	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-headless" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofileud3K2i"
*** You are running in headless mode.
1602674403135	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602674403583	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602674403583	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602674403583	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602674403583	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602674405308	Marionette	INFO	Listening on port 46397
1602674405404	Marionette	WARN	TLS certificate errors will be ignored for this session
1602674411796	Marionette	INFO	Stopped listening on port 46397
JavaScript error: , line 0: uncaught exception: undefined
1602674439459	geckodriver	INFO	Listening on 127.0.0.1:58727
1602674439463	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-headless" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofileJBOpGQ"
*** You are running in headless mode.
1602674439643	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602674439920	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602674439920	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602674439920	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602674439920	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602674440945	Marionette	INFO	Listening on port 34439
1602674440968	Marionette	WARN	TLS certificate errors will be ignored for this session
1602674446341	Marionette	INFO	Stopped listening on port 34439
JavaScript error: , line 0: uncaught exception: undefined
1602674474476	geckodriver	INFO	Listening on 127.0.0.1:55361
1602674474480	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-headless" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofiletaForl"
*** You are running in headless mode.
1602674474665	addons.webextension.doh-rollout@mozilla.org	WARN	Loading extension 'doh-rollout@mozilla.org': Reading manifest: Invalid extension permission: networkStatus
1602674474931	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1602674474931	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
1602674474931	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1602674474931	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1602674476006	Marionette	INFO	Listening on port 40417
1602674476085	Marionette	WARN	TLS certificate errors will be ignored for this session
1602674483662	Marionette	INFO	Stopped listening on port 40417
JavaScript error: , line 0: uncaught exception: undefined
```

The thing that I don't understand is why it doesn't return proper result when using into PHP script and works fine when using it alone?

**Comment by whimboo** [comment]
Can you please test by using the real Firefox binary and not the shell script at `/usr/bin/firefox`. Does that fix your issue? Note that there is also issue #1756, which will be fixed in the next geckodriver release. So feel free to test the geckodriver nightly build from https://github.com/mozilla/geckodriver/issues/1756#issuecomment-673576170.

**Comment by whimboo** [comment]
Also note that the above stacktrace is not a trace log (as requested). You would have to set at least the Firefox preference `marionette.log.level=Trace`.

**Comment by ishan-biztech** [comment]
Hi @whimboo, here's the [geckodriver.log](https://github.com/mozilla/geckodriver/files/5384542/geckodriver.log).

Also can you please explain what you mean by "Can you please test by using the real Firefox binary and not the shell script at `/usr/bin/firefox`."?

This script is working fine in my local machine(Ubuntu 18.10), but not working on the server only when called from PHP script. I've already checked all the permissions.

**Comment by whimboo** [comment]
So the last attached geckodriver.log is from where? Your local machine or from the server? I cannot see a problem in there, except the fact that it is using a kind old version of Firefox (76). You should consider to upgrade to 81.0.

Note that the Javascript error might come from some Firefox component that might already have been fixed in a newer release. Please check that.

**Comment by ishan-biztech** [comment]
Hi @whimboo, the `geckodriver.log` is from the server. And regarding updating the Firefox browser, sure I will do it and update you on the same.

**Comment by ishan-biztech** [comment]
And one thing is that if the problem is of Firefox browser why does it work from command line? The browser version is same there too.

What are your thoughts on that?

**Comment by whimboo** [comment]
I don't know. Please test a more up2date version of Firefox first.

**Comment by ishan-biztech** [comment]
I've updated Firefox version to `80.0.1` but it still does not work. It stops unexpectedly when it reaches `self.driver = webdriver.Firefox(options=options)` line.

I also checked [this stackoverflow thread](https://stackoverflow.com/a/18350012) and set proper permissions too, but it does not work.

**Comment by ishan-biztech** [comment]
There was multiple issues here(only when we run php with web server in our case it was apache),

1. Missing HOME directory, when I used try...except to check if there was any issue. It was throwing "Could not determine HOME directory" error. While searching for this issue I landed on [this](https://stackoverflow.com/a/59127977) stackoverflow thread. So I fixed it by setting HOME environment using `os.environ['HOME'] = '/home/<username>'`

2. The `$PATH` variable was not set for `geckodriver`, it was working because I set `$PATH` var to `/home/<username>/.local/bin`(where I put `geckodriver` file) and exported it into my `~/.zshrc` file. But this was not available when I run it from Apache web server. So to fix this I copied `geckodriver` file from `/home/<username>/.local/bin` to `/usr/local/bin`.

I'm closing this issue now since it is not an issue with `geckodriver` but issue was with my python script. Thanks for the help @whimboo.
